### PR TITLE
fix(optimizer): fix lookup join and agg distribution derive

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Lack of considering `output_indices` for `BatchLookupJoin`.
- Lack of considering `UpstreamHashShard` for `BatchHashAgg`, `BatchSortAgg`, `StreamHashAgg`.
- This bug found by enabling batch lookup join by default. https://github.com/risingwavelabs/risingwave/pull/6818

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/frontend/planner_test/tests/testdata/batch_index_join.yaml [hunk 1]
```diff
@@ -58,3 +58,18 @@
...
+    set rw_batch_enable_lookup_join = true;
+    set query_mode = distributed;
+    create table t (a int, b int);
+    create table t2 (c int, d int, primary key(c));
+    select t2.c, t2.d, count(distinct t.a) from t join t2 on t.a = t2.c group by t2.c, t2.d;
+  batch_plan: |
+    BatchExchange { order: [], dist: Single }
+    └─BatchHashAgg { group_key: [t2.c, t2.d], aggs: [count(t.a)] }
+      └─BatchExchange { order: [], dist: HashShard(t2.c, t2.d) }
+        └─BatchHashAgg { group_key: [t2.c, t2.d, t.a], aggs: [] }
```
Comment (BowenXiao1999) [inline src/frontend/planner_test/tests/testdata/batch_index_join.yaml:72]: Before this pr, we can not derive the group_key here?

**Review by BowenXiao1999 — APPROVED** [review]
LGTM

File: src/frontend/planner_test/tests/testdata/batch_index_join.yaml (hunk 1 above)
Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/batch_index_join.yaml:72]: Before this pr, we will miss a exchange here.

**Comment by BugenZhao** [comment]
Will #6642 be caused by this? 🤔

File: src/frontend/planner_test/tests/testdata/batch_index_join.yaml [hunk 2]
```diff
@@ -58,3 +58,18 @@
     └─BatchLookupJoin { type: Inner, predicate: t.a = idx.c, output: all }
       └─BatchExchange { order: [], dist: UpstreamHashShard(t.a) }
         └─BatchScan { table: t, columns: [t.a, t.b], distribution: SomeShard }
+- name: test index join distribution derive
+  sql: |
+    set rw_batch_enable_lookup_join = true;
+    set query_mode = distributed;
```
Comment (BugenZhao) [inline src/frontend/planner_test/tests/testdata/batch_index_join.yaml:64]: I guess the `batch_plan` must be distributed?

Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/batch_index_join.yaml:64]: Currently, our query execution mode by default is `local`, so we need to set it to `distributed`.

Comment (BugenZhao) [reply src/frontend/planner_test/tests/testdata/batch_index_join.yaml:64]: Seems we always generate distributed plan. For local plan, we use the `local_batch_plan` field.
https://github.com/risingwavelabs/risingwave/blob/672c049254904aff99038366a57e3663aa0052b2/src/frontend/planner_test/src/lib.rs#L468-L479

Comment (chenzl25) [reply src/frontend/planner_test/tests/testdata/batch_index_join.yaml:64]: You are right. I can remove it in this https://github.com/risingwavelabs/risingwave/pull/6818 PR together.
```
set rw_batch_enable_lookup_join = true;
set query_mode = distributed;
```

**Pull request merged**